### ISSUE TITLE
Pyright hangs indefinitely when used with isort SortImports

### ISSUE THREAD

**Issue by floredelasteyrie** [microsoft/pyright#1642]
**Describe the bug**
Pyright hangs indefinitely when the file to analyse contains isort `SortImports`

**To Reproduce**
File `test.py`
```python
from isort import SortImports

with open('some_file.txt', 'r') as handler:
    SortImports(handler, quiet=True)
```

With the following `pyrightconfig.json`:
```json
{
    "useLibraryCodeForTypes": true
}
```

Then `pyright test.py`

**Expected behavior**
I would expect Pyright to analyse the file successfully

**VS Code extension or command-line**
Command-line v1.1.122

**Additional context**
python 3.8.6
isort 4.3.21

**Comment by erictraut** [comment]
I'm not able to reproduce a hang. I just receive an error indicating that `SortImports` is an unknown import symbol. When I search the contents of the package, I can't find `SortImports`.

When you say that Pyright hangs, can you explain what symptom you're seeing? Are you using VS Code? If so, does the entire UI stop responding? I'm wondering if it's possibly some other extension that's causing your hang.

**Comment by erictraut** [comment]
I just re-read your bug report, and I noticed that you're using the command-line version of pyright. Here's what I see when I run pyright on the sample above. I have isort 4.3.21 installed within my venv.

```
Found 1 source file
  /Users/eric/source/test.py:1:19 - error: "SortImports" is unknown import symbol (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos 
Completed in 0.658sec
```

Here are some things to try:
1. Try running pyright on an empty ".py" file. Does it still hang?
2. Try running it on your sample code with the `with` statement commented out (i.e. only the import statement present). Does it still hang?
3. Check the version of node that you have installed. Is it fairly recent (14.x)?

**Comment by floredelasteyrie** [comment]
Thanks for taking a look ! 
Weird that it doesn't hang on your side.. Here are some more details: 
- It doesn't hang if I comment the line `SortImports(handler, quiet=True)`: it's not the import that blocks but the call itself
- It also doesn't hang if I remove the `pyrightconfig.json` above
- I'm using node v15.3.0 / npm 7.6.0
- As a matter of fact i'm also using VS Code, in which I observed the entire file isn't analysed when it contains this line, but I was using Pylance v2021.3.1 which wasn't up to date with the latest Pyright. I just tried with the v2021.3.2 this morning, doesn't seem to analyse it either.

When it hangs in the command-line, here are the logs:
```bash
Loading configuration file at /Users/floredelasteyrie/test-pyright/pyrightconfig.json
Assuming Python version 3.8
Assuming Python platform Darwin
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding .git
stubPath /Users/floredelasteyrie/test-pyright/typings is not a valid directory.
Searching for source files
Found 1 source file
```

**Comment by erictraut** [comment]
Thanks for the additional details. I'm now able to repro the hang. I must have had the wrong version of isort installed previously. I'll debug the problem.

**Comment by erictraut** [comment]
The hang was due to a particularly complex code flow graph in the isort.py file. I've added some heuristics that disable the more advanced type inference techniques under circumstances where performance is likely to be problematic, and this avoids the problem. The fix will be included in the next release of pyright.

**Comment by floredelasteyrie** [comment]
Great, thanks !

**Comment by erictraut** [comment]
This is now addressed in pyright 1.1.123, which I just published. It will also be included in the next release of pylance.
